crates/order-validation/src/banned.rs

```diff
@@ -1,22 +1,128 @@
 use {
-    cached::{Cached, TimedCache},
     contracts::ChainalysisOracle,
     ethcontract::{errors::MethodError, futures::future::join_all, H160},
-    std::{collections::HashSet, sync::Mutex},
+    std::{
+        collections::{HashMap, HashSet},
+        ops::Div,
+        sync::Arc,
+        time::{Duration, Instant},
+    },
+    tokio::sync::RwLock,
 };
 
 /// A list of banned users and an optional registry that can be checked onchain.
 pub struct Users {
     list: HashSet<H160>,
-    onchain: Option<Onchain>,
+    onchain: Option<Arc<Onchain>>,
+}
+
+#[derive(Clone)]
+struct UserMetadata {
+    is_banned: bool,
+    last_updated: Instant,
+}
+
+impl UserMetadata {
+    pub fn with_banned(mut self, banned: bool) -> Self {
+        self.is_banned = banned;
+        self
+    }
+
+    pub fn with_last_updated(mut self, last_updated: Instant) -> Self {
+        self.last_updated = last_updated;
+        self
+    }
 }
 
 struct Onchain {
     contract: ChainalysisOracle,
-    cache: Mutex<TimedCache<H160, bool>>,
+    cache: RwLock<HashMap<H160, UserMetadata>>,
 }
 
-const TTL: u64 = 60 * 60; // 1 hour
+impl Onchain {
+    pub fn new(contract: ChainalysisOracle) -> Arc<Self> {
+        let onchain = Arc::new(Self {
+            contract,
+            cache: Default::default(),
+        });
+
+        onchain.clone().spawn_maintenance_task();
+
+        onchain
+    }
+
+    /// Spawns a background task that periodically checks the cache for expired
+    /// entries and re-run checks for them.
+    ///
+    /// Removes from the cache entries that are expired and not limit order
+    /// participants and not part of the current auction. This is made to keep
+    /// addresses of long-lasting limit orders in the cache, which is the vast
+    /// majority. And addresses that open market orders which get settled pretty
+    /// quickly should get flushed out again.
+    fn spawn_maintenance_task(self: Arc<Self>) {
+        let cache_expiry = Duration::from_secs(60 * 60);
+        let maintenance_timeout = cache_expiry.div(10).max(Duration::from_secs(60));
```

Comment:
Why do we need the two concepts of "cache expiry" and "maintenance timeout" here? Since we are now using a HashMap instead of a TimedCache it feels pointless to differentiate between "when to update" and "when object expire" (in theory even expired entries would be returned with this implementation).

Reply:
Different elements get inserted into the cache at different time points. `cache_expiry` is required to understand if a particular item has expired. Not sure how should that work without this config.

`maintenance_timeout` just defines how often the task should be executed regardless of the time spent on each iteration.

Not sure I understand how this should work with only one of the configs.

Reply:
Like you could just update all values regardless of how long they were in there once per hour. But ok it doesn't matter much and might distribute things more evenly this way.